src/ipa_pc/data_structures.rs

```diff
@@ -51,83 +107,89 @@ impl<G: AffineCurve> PCUniversalParams for UniversalParams<G> {
 /// polynomial.
 #[derive(Derivative)]
 #[derivative(
-    Default(bound = ""),
     Hash(bound = ""),
     Clone(bound = ""),
     Debug(bound = ""),
     Eq(bound = ""),
     PartialEq(bound = "")
 )]
 #[derive(CanonicalSerialize, CanonicalDeserialize)]
-pub struct CommitterKey<G: AffineCurve> {
-    /// The key used to commit to polynomials.
-    pub comm_key: Vec<G>,
+pub struct CommitterKey<G: Curve> {
+    /// The base point vector used for the coefficients of the polynomial.
+    pub comm_key: Vec<G::AffineRep>,
 
-    /// A random group generator.
+    /// The base point used by the opening proof to reduce to a non-blinded
+    /// opening assertion.
     pub h: G,
 
-    /// A random group generator that is to be used to make
-    /// a commitment hiding.
+    /// The base point that is to be used for hiding.
     pub s: G,
 
     /// The maximum degree supported by the parameters
     /// this key was derived from.
     pub max_degree: usize,
 
-    /// H(max_degree_comm_key, h, s, max_degree)
+    /// A fingerprint of the CommitterKey, H(max_degree_comm_key, h, s, max_degree).
+    /// Used for reference.
     pub hash: Vec<u8>,
 }
 
-impl<G: AffineCurve> CommitterKey<G> {
+impl<G: Curve> CommitterKey<G> {
     /// Scale key for testing purpose
     pub fn scale(&mut self, scaling_scalar: G::ScalarField) {
-        self.h = self.h.mul(scaling_scalar).into_affine();
+        self.h = self.h.mul(&scaling_scalar);
     }
 }
 
-impl<G: AffineCurve> SemanticallyValid for CommitterKey<G> {
+impl<G: Curve> SemanticallyValid for CommitterKey<G> {
     // Technically this function is redundant, since the keys are generated
     // through a deterministic procedure starting from a public string.
     fn is_valid(&self) -> bool {
-        self.comm_key.is_valid()
+        G::batch_from_affine(&self.comm_key).is_valid()
```

Comment:
Why SemanticallyValid is not implemented for Curve ?

Reply:
Actually SemanticallyValid is implemented for Curve and not implemented for Curve::AffineRep. The comm_key is saved as Vec<G::AffineRep> because it is more effective in terms of memory usage and also it is used in MSM where our most efficient implementation is based on Affine representation.

Our approach was to use AffineRep only for efficient arithmetic, and implement as little logic for it as possible, to stay "projective" everywhere when possible.